# WEEK-01
Last week, we learned how to code in Python and extract data from websites.

Data is like fuel for a car—without it, there is no AI. But having data alone isn’t enough; we need to understand, clean, and prepare it before training our models.

This week, we’ll explore new tools and concepts, so be patient and make the most of this learning experience! 🚀
and remeber


**5 minutes of reading can save you 6 hours of debugging**


Please use the dataset included in the subject. You can find it in the [GitHub repository](https://github.com/1337-Artificial-Intelligence/Entry-Level-ML-Engineer-Bootcamp/tree/main/week01/Housing.csv).

Good luck! 🤞



---



## Exercise-01: Load and Inspect the Dataset

The goal of this exercise is to load a dataset and understand its structure.

We encourage you to use the Pandas library for this task.
To get started, explore the Pandas library by checking its [official documentation:](https://pandas.pydata.org/pandas-docs/stable/user_guide/index.html)

### **Instructions**
1. Load the dataset into a Pandas DataFrame.
2. Print the number of rows in the dataset.


### **Open Question**

>🙋*What is a DataFrame, and why do we need it?*


In [4]:
import pandas as pd
df = pd.read_csv("./Housing.csv")

print("Number of rows in the dataset:", len(df.index))
print("Number of rows in the dataset:", df.shape[0])
print("Number of rows in the dataset:", df[df.columns[0]].count())#number of non NaN values hmmm

Number of rows in the dataset: 559
Number of rows in the dataset: 559
Number of rows in the dataset: 558


The output should be:

`Number of rows in the dataset: 559`

---
Now that we have successfully loaded the dataset and counted the number of rows, let's display it.

However, when working with large datasets, it's not practical to display everything at once. Instead, let's display only the first 5 rows to get a quick overview! 🚀

In [8]:
df.head()

price    area  bedrooms  bathrooms  stories mainroad guestroom  \
0  13300000.0  7420.0         4          2      3.0      yes        no   
1  12250000.0  8960.0         4          4      4.0      yes        no   
2  12250000.0  9960.0         3          2      2.0      yes        no   
3  12215000.0  7500.0         4          2      2.0      yes        no   
4  11410000.0  7420.0         4          1      2.0      yes       yes   

  basement hotwaterheating airconditioning  parking prefarea furnishingstatus  
0       no              no             yes      2.0      yes        furnished  
1       no              no             yes      3.0       no        furnished  
2      yes              no              no      2.0      yes   semi-furnished  
3      yes              no             yes      3.0      yes        furnished  
4      yes              no             yes      2.0       no        furnished

The output should be like this



```
	price	area	bedrooms	bathrooms	stories	mainroad	guestroom	basement	hotwaterheating	airconditioning	parking	prefarea	furnishingstatus
0	13300000.0	7420.0	4	2	3.0	yes	no	no	no	yes	2.0	yes	furnished
1	12250000.0	8960.0	4	4	4.0	yes	no	no	no	yes	3.0	no	furnished
2	12250000.0	9960.0	3	2	2.0	yes	no	yes	no	no	2.0	yes	semi-furnished
3	12215000.0	7500.0	4	2	2.0	yes	no	yes	no	yes	3.0	yes	furnished
4	11410000.0	7420.0	4	1	2.0	yes	yes	yes	no	yes	2.0	no	furnished

```



---
Now that we've loaded the dataset and displayed the first few rows, let's dive deeper and understand its structure.
To do this, use a function that provides essential details about the dataset, such as:

- The number of non-null values in each column
- The data types of each column
- The total number of entries



In [10]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 559 entries, 0 to 558
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   price             558 non-null    float64
 1   area              558 non-null    float64
 2   bedrooms          559 non-null    int64  
 3   bathrooms         559 non-null    int64  
 4   stories           557 non-null    float64
 5   mainroad          559 non-null    object 
 6   guestroom         559 non-null    object 
 7   basement          559 non-null    object 
 8   hotwaterheating   559 non-null    object 
 9   airconditioning   557 non-null    object 
 10  parking           558 non-null    float64
 11  prefarea          558 non-null    object 
 12  furnishingstatus  557 non-null    object 
dtypes: float64(4), int64(2), object(7)
memory usage: 56.9+ KB


Output format

```
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 559 entries, 0 to 558
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   price             558 non-null    float64
 1   area              558 non-null    float64
 2   bedrooms          559 non-null    int64  
 3   bathrooms         559 non-null    int64  
 4   stories           557 non-null    float64
 5   mainroad          559 non-null    object
 6   guestroom         559 non-null    object
 7   basement          559 non-null    object
 8   hotwaterheating   559 non-null    object
 9   airconditioning   557 non-null    object
 10  parking           558 non-null    float64
 11  prefarea          558 non-null    object
 12  furnishingstatus  557 non-null    object
dtypes: float64(4), int64(2), object(7)
memory usage: 56.9+ KB
```



---

Before we start processing the dataset, it's important to check if there are any missing values. Missing data can affect model performance and lead to inaccurate results, so identifying and handling them is a key step in data preparation.

Try using a method that allows you to see how many missing values exist in each column.


In [17]:
df.isnull().sum()

price               1
area                1
bedrooms            0
bathrooms           0
stories             2
mainroad            0
guestroom           0
basement            0
hotwaterheating     0
airconditioning     2
parking             1
prefarea            1
furnishingstatus    2
dtype: int64

Output format


```
price               1
area                1
bedrooms            0
bathrooms           0
stories             2
mainroad            0
guestroom           0
basement            0
hotwaterheating     0
airconditioning     2
parking             1
prefarea            1
furnishingstatus    2
dtype: int64
```



---
# Exercise 2: Data Cleaning

Now that we've identified missing values and inspected our dataset, it's time to clean the data!

In this exercise, you'll focus on:

Removing or handling missing values
Identifying and removing duplicate rows
Cleaning data is a crucial step to ensure our model is trained on high-quality, reliable data.
### 🙋 Open Question
> Why is it important to remove duplicate rows in a dataset?

In [ ]:
# add your code here

The output should be


```

price	area	bedrooms	bathrooms	stories	mainroad	guestroom	basement	hotwaterheating	airconditioning	parking	prefarea	furnishingstatus
0	13300000.0	7420.0	4	2	3.0	yes	no	no	no	yes	2.0	yes	furnished
1	12250000.0	8960.0	4	4	4.0	yes	no	no	no	yes	3.0	no	furnished
2	12250000.0	9960.0	3	2	2.0	yes	no	yes	no	no	2.0	yes	semi-furnished
3	12215000.0	7500.0	4	2	2.0	yes	no	yes	no	yes	3.0	yes	furnished
4	11410000.0	7420.0	4	1	2.0	yes	yes	yes	no	yes	2.0	no	furnished
...	...	...	...	...	...	...	...	...	...	...	...	...	...
547	1820000.0	3000.0	2	1	1.0	yes	no	yes	no	no	2.0	no	unfurnished
548	1767150.0	2400.0	3	1	1.0	no	no	no	no	no	0.0	no	semi-furnished
549	1750000.0	3620.0	2	1	1.0	yes	no	no	no	no	0.0	no	unfurnished
550	1750000.0	2910.0	3	1	1.0	no	no	no	no	no	0.0	no	furnished
551	1750000.0	3850.0	3	1	2.0	yes	no	no	no	no	0.0	no	unfurnished
544 rows × 13 columns```



---
# Exercice 3 : Descriptive Statistics

Now that we have cleaned our dataset, it's time to explore and understand it better using **descriptive statistics**. These statistics help summarize the data and give us insights into its distribution, central tendency, and variability.  

### **Your Tasks**  
1. Use Pandas to calculate basic statistics such as:  
   - Mean, median, and standard deviation of numerical columns.  
   - Minimum and maximum values.  
   - Count of non-null values in each column.
2. Check the distribution of numerical features


3.  Show unique values for categorical variables.


### 🙋 Open Question
> Why do we need descriptive statistics when working with data?

🚀 **Let’s analyze our dataset!**

###1. Calculate basic statistics

In [ ]:
# add your code here

*Your* output should be in this format:


```
Basic Statistics:
              price          area    bedrooms   bathrooms     stories  \
count  5.580000e+02    558.000000  559.000000  559.000000  557.000000   
mean   4.777083e+06   5135.637993    2.967800    1.288014    1.811490   
std    1.885783e+06   2157.286148    0.734969    0.502020    0.872446   
min    1.750000e+06   1650.000000    1.000000    1.000000    1.000000   
25%    3.438750e+06   3600.000000    2.500000    1.000000    1.000000   
50%    4.340000e+06   4515.000000    3.000000    1.000000    2.000000   
75%    5.808250e+06   6360.000000    3.000000    2.000000    2.000000   
max    1.330000e+07  16200.000000    6.000000    4.000000    4.000000   

          parking  
count  558.000000  
mean     0.691756  
std      0.858982  
min      0.000000  
25%      0.000000  
50%      0.000000  
75%      1.000000  
max      3.000000  

```



###2.Check the distribution of numerical features using:  


In [ ]:
# add your code here

Your output should be in this format:

```
Distribution of numerical features

Value Counts for mainroad:
mainroad
yes    481
no      78
Name: count, dtype: int64

Value Counts for guestroom:
guestroom
no     460
yes     99
Name: count, dtype: int64

Value Counts for basement:
basement
no     365
yes    194
Name: count, dtype: int64

Value Counts for hotwaterheating:
hotwaterheating
no     533
yes     26
Name: count, dtype: int64

Value Counts for airconditioning:
airconditioning
no     382
yes    175
Name: count, dtype: int64

Value Counts for prefarea:
prefarea
no     425
yes    133
Name: count, dtype: int64

Value Counts for furnishingstatus:
furnishingstatus
semi-furnished    230
unfurnished       183
furnished         144
Name: count, dtype: int64
```



###3.Show unique values for categorical variables.


In [ ]:
# add your code here

Output format


```
Unique values in 'mainroad':
['yes' 'no']

Unique values in 'guestroom':
['no' 'yes']

Unique values in 'basement':
['no' 'yes']

Unique values in 'hotwaterheating':
['no' 'yes']

Unique values in 'airconditioning':
['yes' 'no' nan]

Unique values in 'prefarea':
['yes' 'no' nan]

Unique values in 'furnishingstatus':
['furnished' 'semi-furnished' 'unfurnished' nan]
```



#Exercise 4: Univariate Analysis**  

In this exercise, you'll analyze the distribution of each feature using visualization techniques. Understanding how your data is distributed is crucial for identifying patterns, outliers, and potential preprocessing steps.  

### **Tasks**  
1️⃣ **Plot histograms** for numerical variables to see their distribution.  
2️⃣ **Display boxplots** to detect outliers.  
3️⃣ **Plot pie charts** for categorical variables to understand their proportions.  

💡 **Hint**: Use **Matplotlib** and **Seaborn** for plotting. You can explore their documentation here:  
- [Matplotlib Docs](https://matplotlib.org/)  
- [Seaborn Docs](https://seaborn.pydata.org/)  

Happy coding! 🚀

In [ ]:
# add your code here

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

**Learning takes time! If things feel challenging, that's completely normal. Be patient, keep exploring, and trust the process—you’re making progress!**

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

---
#Exercise 5: Bivariate Analysis
Now that we have explored individual features, let’s analyze the relationships between two variables. This will help us understand how different features interact with each other.

Tasks

1️⃣ Scatter plots – Visualize relationships between numerical variables.

2️⃣ Correlation heatmap – Identify correlations between numerical features.

3️⃣ Boxplots – Compare categorical and numerical variables.

💡 Hint: Use Matplotlib and Seaborn for visualization.

Scatter plots help detect trends or patterns.
The heatmap shows how strongly two numerical variables are related.
Boxplots help compare distributions across categories.
Let’s dive in! 🚀

In [ ]:
# add your code here

---
#Exercise 6: Multivariate Analysis

Now that we've explored relationships between two variables, let's analyze interactions among multiple variables simultaneously. This will help us uncover deeper patterns in the data.

Tasks

1️⃣ Pairplot – Visualize relationships between multiple numerical variables.

2️⃣ Multivariate correlation heatmap – Identify correlations across multiple features.

3️⃣ Grouped boxplots – Compare distributions of a numerical variable across multiple categorical variables.

💡 Hint:

Use Seaborn's pairplot to explore all pairwise relationships.

The heatmap helps identify groups of strongly correlated variables.

Grouped boxplots provide insights into variations across different categories.

In [ ]:
# add your code here

---
#Exercise 7: Making new features

Feature engineering is a powerful way to improve model performance by adding meaningful information that the machine can't detect on its own.  

### **Tasks**  
1️⃣ Create **one new feature** based on existing columns.  
   - You can **combine** or **split** columns.  
   - Extract meaningful insights (e.g., **datetime → morning/evening**, **weekend/weekday**, **text length**, etc.).  

2️⃣ Add these new features to your dataset.  


3️⃣ Display the first few rows to check the results.  



### 🙋 Open Question

> How do new features impact machine learning models? Can you think of a real-world example where feature engineering made a big difference?

Get creative and enhance your dataset! 🚀

In [ ]:
#add your code here

# **Exercise 8: Preparing Data for Machine Learning**  

To make our dataset machine learning-friendly, we need to **convert categorical variables into numerical values** and **scale numerical features** for better model performance.  

### **Tasks**  
1️⃣ **Convert ordinal categorical variables** into numerical values using **Label Encoding**.  
   - Example: `("yes","no") → (1,0)`  
   - Example: `("very unsatisfied", "unsatisfied", "neutral", "satisfied", "very satisfied") → (0,1,2,3,4)`  

2️⃣ **Apply One-Hot Encoding** for **nominal categorical variables** (where order doesn't matter).  

3️⃣ **Scale numerical variables** using any scaler (MinMaxScaler, StandardScaler, etc.).  
   - Choose the best scaler for your dataset and **justify your choice**.  

💡 **Hint**:  
- Use **Label Encoding** for categories with an inherent order.  
- Use **OneHotEncoding** for categories without order (e.g., colors, cities).  
- Scaling helps models **converge faster** and improves **performance**.  

### 🙋 Open Question
> Why do we need to scale numerical variables? Can you think of a case where scaling isn't necessary?

Let's transform our dataset into ML-ready data! 🚀



---

## 🎉 Congratulations! You've Completed Week01!  

Great job on making it through week 1 of the bootcamp! 🚀 You've learned how to load, clean, and explore data—essential skills for any ML engineer.  

Data preprocessing is a crucial step in any machine learning project, and mastering it will set you up for success in the coming weeks.  

🔍 **What’s Next?**  
Take some time to review what you’ve learned and apply it to different datasets. If you have any questions or got stuck along the way, don’t hesitate to ask in our community!  

Stay curious, keep experimenting, and get ready for Week 2! 💡🔥